File: go/bsb/observable.go
package bsb

// Observable provides unified access to logging, metrics, and tracing.
// Observable instances are immutable -- attribute methods return new instances.
type Observable interface {
	// Trace returns the distributed trace context.
	Trace() DTrace

	// TraceID returns the 32-char hex trace ID.
	TraceID() string

	// SpanID returns the 16-char hex span ID.
	SpanID() string

	// Resource returns the OpenTelemetry resource context.
	Resource() ResourceContext

	// Attributes returns the current attribute set.
	Attributes() map[string]any

	// Log returns the structured logger.
	Log() ObservableLog

	// Metrics returns the metrics factory.
	Metrics() ObservableMetrics

	// StartSpan creates a child span with a new span ID.
	StartSpan(name string, attributes ...map[string]any) Observable

	// SetAttribute returns a new Observable with the added attribute.
	SetAttribute(key string, value any) Observable

	// SetAttributes returns a new Observable with the added attributes.
	SetAttributes(attrs map[string]any) Observable

	// Error records an error to both logs and the active span.
	Error(err error, attributes ...map[string]any)

	// End finalizes the current span.
	End(attributes ...map[string]any)
}

// ObservableLog provides structured logging methods.
type ObservableLog interface {
	Debug(message string, meta ...map[string]any)
	Info(message string, meta ...map[string]any)
	Warn(message string, meta ...map[string]any)
	Error(message string, meta ...map[string]any)
}

// ObservableMetrics provides metric creation methods.
type ObservableMetrics interface {
	Counter(name, description, help string) *Counter
	Gauge(name, description, help string) *Gauge
	Histogram(name, description, help string, boundaries ...[]float64) *Histogram
	Timer() Timer
}

// ObservableBackendInterface is the internal contract for observable backends.
// Observable plugins and the framework implement this to receive log/metric/trace calls.
type ObservableBackendInterface interface {
	// Logging
	Debug(trace DTrace, pluginName, message string, meta map[string]any)
	Info(trace DTrace, pluginName, message string, meta map[string]any)
	Warn(trace DTrace, pluginName, message string, meta map[string]any)
	LogError(trace DTrace, pluginName, message string, meta map[string]any)

	// Metrics
	CreateCounter(pluginName, name, description, help string) *Counter
	CreateGauge(pluginName, name, description, help string) *Gauge
	CreateHistogram(pluginName, name, description, help string, boundaries []float64) *Histogram
	CreateTimer() Timer

	// Tracing
	StartSpan(trace DTrace, pluginName, name string, attributes map[string]any) (DTrace, string)
	EndSpan(trace DTrace, pluginName, spanID string, attributes map[string]any)
	ErrorSpan(trace DTrace, pluginName, spanID string, err error, attributes map[string]any)
}

File: dotnet/BSB/Interfaces/Observable.cs
namespace BSB.Interfaces;

/// <summary>
/// Structured log entry metadata. Key-value pairs attached to a log message.
/// </summary>
public class LogMeta : Dictionary<string, object?>
{
    /// <summary>
    /// Create an empty metadata container.
    /// </summary>
    public LogMeta() { }

    /// <summary>
    /// Create metadata from an existing dictionary.
    /// </summary>
    public LogMeta(IDictionary<string, object?> dict) : base(dict) { }
}

/// <summary>
/// Logging interface for observable spans. All log calls are attached to the current span context.
/// </summary>
public interface IObservableLog
{
    /// <summary>
    /// Log a debug-level message.
    /// </summary>
    void Debug(string message, LogMeta? meta = null);

    /// <summary>
    /// Log an informational message.
    /// </summary>
    void Info(string message, LogMeta? meta = null);

    /// <summary>
    /// Log a warning message.
    /// </summary>
    void Warn(string message, LogMeta? meta = null);

    /// <summary>
    /// Log an error message.
    /// </summary>
    void Error(string message, LogMeta? meta = null);

    /// <summary>
    /// Log an exception with an optional message.
    /// </summary>
    void Error(Exception ex, string? message = null, LogMeta? meta = null);
}

/// <summary>
/// Metrics factory interface for observable spans.
/// Creates metric instruments scoped to the current service context.
/// </summary>
public interface IObservableMetrics
{
    /// <summary>
    /// Create a monotonically increasing counter.
    /// </summary>
    ICounter Counter(string name, string description, string unit);

    /// <summary>
    /// Create a gauge that can go up and down.
    /// </summary>
    IGauge Gauge(string name, string description, string unit);

    /// <summary>
    /// Create a histogram for recording value distributions.
    /// </summary>
    IHistogram Histogram(string name, string description, string unit);

    /// <summary>
    /// Create a timer that measures elapsed duration.
    /// </summary>
    ITimer Timer();
}

/// <summary>
/// OpenTelemetry-compatible resource context for identifying the service instance.
/// </summary>
public class ResourceContext
{
    /// <summary>
    /// Logical name of the service (e.g. "my-api").
    /// </summary>
    public string ServiceName { get; init; } = "";

    /// <summary>
    /// Version of the service (e.g. "1.2.3").
    /// </summary>
    public string ServiceVersion { get; init; } = "";

    /// <summary>
    /// Unique identifier for this service instance.
    /// </summary>
    public string ServiceInstanceId { get; init; } = "";

    /// <summary>
    /// Deployment environment (e.g. "production", "staging").
    /// </summary>
    public string DeploymentEnvironment { get; init; } = "";

    /// <summary>
    /// Optional deployment region (e.g. "us-east-1").
    /// </summary>
    public string? DeploymentRegion { get; init; }

    /// <summary>
    /// Convert to OpenTelemetry resource attribute dictionary.
    /// </summary>
    public Dictionary<string, string> ToAttributes()
    {
        var attrs = new Dictionary<string, string>
        {
            ["service.name"] = ServiceName,
            ["service.version"] = ServiceVersion,
            ["service.instance.id"] = ServiceInstanceId,
            ["deployment.environment"] = DeploymentEnvironment,
        };
        if (DeploymentRegion is not null)
            attrs["deployment.region"] = DeploymentRegion;
        return attrs;
    }
}

/// <summary>
/// Core observability interface - unified logging, metrics, and distributed tracing.
/// All plugin lifecycle methods receive an <see cref="IObservable"/> for structured telemetry.
/// </summary>
public interface IObservable
{
    /// <summary>
    /// The distributed trace context for this span.
    /// </summary>
    DTrace Trace { get; }

    /// <summary>
    /// Shortcut for <c>Trace.TraceId</c>.
    /// </summary>
    string TraceId { get; }

    /// <summary>
    /// Shortcut for <c>Trace.SpanId</c>.
    /// </summary>
    string SpanId { get; }

    /// <summary>
    /// Resource context identifying the service instance.
    /// </summary>
    ResourceContext Resource { get; }

    /// <summary>
    /// Attributes set on this span.
    /// </summary>
    IReadOnlyDictionary<string, object?> Attributes { get; }

    /// <summary>
    /// Structured logging interface scoped to this span.
    /// </summary>
    IObservableLog Log { get; }

    /// <summary>
    /// Metrics factory scoped to this service context.
    /// </summary>
    IObservableMetrics Metrics { get; }

    /// <summary>
    /// Create a child span within the current trace.
    /// </summary>
    IObservable StartSpan(string name, Dictionary<string, object?>? attributes = null);

    /// <summary>
    /// Set a single attribute on this span.
    /// </summary>
    IObservable SetAttribute(string key, object? value);

    /// <summary>
    /// Set multiple attributes on this span.
    /// </summary>
    IObservable SetAttributes(Dictionary<string, object?> attributes);

    /// <summary>
    /// Record an error on this span.
    /// </summary>
    void Error(Exception error, Dictionary<string, object?>? attributes = null);

    /// <summary>
    /// End this span, optionally adding final attributes.
    /// </summary>
    void End(Dictionary<string, object?>? attributes = null);
}
